pkg/pool-weighted/contracts/smart/InvestmentPool.sol

```diff
@@ -0,0 +1,271 @@
+    {
+        uint256 numTokens = tokens.length;
+        InputHelpers.ensureInputLengthMatch(numTokens, normalizedWeights.length);
+
+        _totalTokens = numTokens;
+
+        // I'm time-traveling a bit here - storing the weights in a form where they can be changed
+        uint256 currentTime = block.timestamp;
+        _startGradualWeightChange(currentTime, currentTime, normalizedWeights, normalizedWeights, tokens);
+    }
+
+    function _getMaxTokens() internal pure virtual override returns (uint256) {
+        return _MAX_WEIGHTED_TOKENS;
+    }
+
+    function _getTotalTokens() internal view virtual override returns (uint256) {
+        return _totalTokens;
+    }
+
+    function _scalingFactor(IERC20 token) internal view virtual override returns (uint256) {
+        bytes32 tokenData = _poolState[token];
+
+        // A valid token can't be zero (must have non-zero weights)
+        if (tokenData == 0) {
+            _revert(Errors.INVALID_TOKEN);
+        }
+
+        return _computeScalingFactor(tokenData);
+    }
+
+    function _scalingFactors() internal view virtual override returns (uint256[] memory scalingFactors) {
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
+        uint256 numTokens = tokens.length;
+
+        scalingFactors = new uint256[](numTokens);
+
+        for (uint256 i = 0; i < numTokens; i++) {
+            scalingFactors[i] = _computeScalingFactor(_poolState[tokens[i]]);
+        }
+    }
+
+    function _getNormalizedWeight(IERC20 token) internal view override returns (uint256) {
+        bytes32 tokenData = _poolState[token];
+
+        // A valid token can't be zero (must have non-zero weights)
+        if (tokenData == 0) {
+            _revert(Errors.INVALID_TOKEN);
+        }
+
+        uint256 startWeight = tokenData.decodeUint64(_START_WEIGHT_OFFSET).uncompress64();
+        uint256 endWeight = tokenData.decodeUint32(_END_WEIGHT_OFFSET).uncompress32();
+
+        uint256 pctProgress = _calculateWeightChangeProgress();
+
+        return _interpolateWeight(startWeight, endWeight, pctProgress);
+    }
+
+    function _getNormalizedWeights() internal view override returns (uint256[] memory normalizedWeights) {
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
+        uint256 numTokens = tokens.length;
+
+        normalizedWeights = new uint256[](numTokens);
+
+        uint256 pctProgress = _calculateWeightChangeProgress();
+
+        for (uint256 i = 0; i < numTokens; i++) {
+            bytes32 tokenData = _poolState[tokens[i]];
+
+            uint256 startWeight = tokenData.decodeUint64(_START_WEIGHT_OFFSET).uncompress64();
+            uint256 endWeight = tokenData.decodeUint32(_END_WEIGHT_OFFSET).uncompress32();
+
+            normalizedWeights[i] = _interpolateWeight(startWeight, endWeight, pctProgress);
+        }
+    }
+
+    function _getNormalizedWeightsAndMaxWeightIndex() internal view override returns (uint256[] memory, uint256) {
```

Comment:
Why is this function ever so slightly different from the one in the LBP? It has the same outcome, but I just don't understand why the conventions are different. I think we should try to be consistent about this stuff.

Reply:
No reason - just wrote it independently. I've made it identical now.